### ISSUE TITLE
performance

### ISSUE THREAD

**Issue by johnrobertallan** [rexm/Handlebars.Net#56]
have you ever run this library through any kind of performance suite? we are trying to render a 100x loop of a template and it seems to run very slowly. i'm not expecting this ticket would resolve our issue, just wondering how extensive the testing has been and if there are any results you can share? thanks very much.

**Comment by rexm** [comment]
Yes, I have done some benchmarks and performance analysis.

Could you share the code? There are some gotchas that aren't always obvious.

**Comment by rexm** [comment]
I don't have them available ATM but I will try to find them.

**Comment by rexm** [comment]
Here are some I found:

Average Timings over 10k runs on a large (long-string), basic template. Timing is in **micro**seconds:

Handlebars against POCO (using reflection): 9,047
Handlebars against Expando (dynamic property resolution): 20,932
Handlebars against JObject (ICustomTypeDescriptor resolution): 185,031

This is one gotcha - when we pass in a JObject as part of the model, Handlebars does not know about JObjects but it does see that the model implements ICustomTypeDescriptor. Based on some tests I ran, it seems that JObject's implementation of ICustomTypeDescriptor is very slow, but I am not sure why.

The other thing I'll mention that's probably obvious but still better to say - is that the parsing, lexing and compiling are very expensive and intended to be one-time, so if you happen to be doing that inside the loop, it will probably be noticeable.

**Comment by johnrobertallan** [comment]
thanks very much for the help. our .NET engineer is going to post some code samples soon. your tests seem very fast!

**Comment by gracehlin** [comment]
Hi @rexm , I work with @johnrobertallan and thanks so much for helping us out!

I'm going to look into the points that you mentioned. In the meanwhile I'm including our code snippets here, if you spot anything fishy in there we'd really appreciate the help!

**C# code that binds data with template and generates the final markup**:
```
string content = GetTemplate(templateName);
var template = handlebars.Compile(content);
var text = template(data);
```
Our performance profiler shows that the `template(data)` took the most amount of time. 
The `data` that we pass in is a list of C# objects: `IEnumerable<ShareFileReblastViewModel>`

**ShareFileReblastViewModel object**: http://pastebin.com/2ZusDLV2
**Handlebars HTML template**: http://pastebin.com/5kX6L8AJ

Right now I'm suspecting that it is because our C# objects are very bloated (as you can see in the first link above) with quite a few complex getters and public methods, which caused the property resolution to take a long time. Would you think this might be a reason?

Note that we also use Handlebars on the frontend with the exact same HTML template and JSON objects, but generating the markup is much faster on the frontend compared to doing it in the .NET backend. 

Thanks so much again!

**Comment by johnrobertallan** [comment]
our loop is actually in the template as an #each. could this in any way be less performant than doing the looping in code?

**Comment by rexm** [comment]
@johnrobertallan the `#each` helper is compiled down to a `foreach`, so you should see no difference there.

One question - we've used words like "slow" and "faster", but can you share the timings you're seeing?

@gracehlin few thoughts here... you said generating the markup on the front-end is much faster. Am I correct in assuming you're using a JsonSerializer on your `IEnumerable<ShareFileReblastViewModel>` and passing the resulting JSON on the client into the HandlebarsJS library. Have you measured how long the serialization takes before sending the JSON to the browser? That would help zero in on your data structure being expensive to traverse, or exclude that as a possibility.

Reviewing the template, it's mostly comprised of basic token replacement, which is generally pretty quick, and a lot of calls to LocalizeResource - do you have any timings on that method? It's being called something like 20 times per iteration. How is the implementation of that helper different from the equivalent helper on the JavaScript side?

**Comment by gracehlin** [comment]
Thanks @rexm!

When I was running it locally on debug, it took about 3 seconds for the markup to be generated, on very little data (~10 ShareFileReblastViewModel objects)... which is a lot slower than the microseconds data that you had there.

To take a step back this is what currently happens when our page loads:
1. Backend retrieves data (`IEnumerable<ShareFileReblastViewModel>`) from datastore
2. Backend calls the Handlebars.NET library to bind data to template. This step takes around 3 seconds total and 98% of it is spent on this call: `template(data)`
3. Render generated markup in the view

Then, in order for our frontend to update the page later, frontend also makes a separate API call to grab our data (`IEnumerable<ShareFileReblastViewModel>`) in JSON format, and binds it to the exact same template using HandlebarsJS. This operation is completely on the frontend and is significant much faster, taking under 1 second. We just want to figure out why Step 2 above, which deals with the same data and same template but just that it's in .NET, would take much longer.

`LocalizeResource` is a custom helper method that we wrote which grabs the localized message string from our resource (.resx) file. This method executes at about 10 nanoseconds, so that would be about 200 nanoseconds (or 0.2 microseconds) per iteration. The resource file is also available to the frontend so the Javascript implementation is quite similar.

I also noticed that when I bind the template in Handlebars.NET, there were a lot of exceptions thrown/caught inside the Handlebars library:

```
InvalidOperationException: Template referenced property name that does not exist.
at Handlebars.Compiler.PathBinder.AccessMember(Object instance, String memberName)
```

```
NullReferenceException
at System.Object.GetType()
at Handlebars.Compiler.PathBinder.AccessMember(Object instance, String memberName)
```

Since .NET exceptions are quite expensive they might be part of the cause. We're investigating which properties caused this.

Thank you so much again for your help Rex!

**Comment by rexm** [comment]
Ah, good find. I'll also look into that. It's entirely possible mine or someone else's laziness overcame good design in some part of the AccessMember logic...

**Comment by rexm** [comment]
Only question I have that isn't explicitly answered - and may be nothing, since I don't know your system internals - is that `IEnumerable<>` to me says "assume it's lazy-loaded", so whoever the first consumer of that enumerable is will actually take the hit of hydrating it. But like I said, I don't know your system so this may not actually be happening. I'll look at the exception thing - that is concerning.

**Comment by rexm** [comment]
OK, yes - [this line](https://github.com/rexm/Handlebars.Net/blob/master/source/Handlebars/Compiler/Translation/Expression/PathBinder.cs#L201) will actually get swallowed and replaced by `UndefinedBindingResult`.

In hindsight this wasn't really a very good coding decision, since the handlebarsjs library treats missing properties as a perfectly acceptable case - not exceptional, so we shouldn't be throwing an exception.

It shouldn't be a lot of work to refactor `AccessMember` and `ResolvePath` to only throw in truly exceptional cases (which should bubble out) and directly return `UndefinedBindingResult` for the rest.

**Comment by gracehlin** [comment]
Hi @rexm, 

You made a very good point regarding `IEnumerable<>`. I don't think we need to worry too much for this case though because even though the code contract says `IEnumerable<T>` (to keep it as generic as possible), under the hood the datastore returns a concrete `List<T>` typed as `IEnumerable<T>`. 

The good news is we have identified the problematic properties in our template that was throwing the `AccessMember` off - after resolving those, the performance has improved drastically :D Binding to 100 objects now takes around 100ms. We're going to continue finetuning our code but Handlebars.NET is no longer the performance bottleneck for us. Thanks so much for your help so far, your prompt replies have made it a lot easier for us!! :thumbsup:

**Comment by johnrobertallan** [comment]
thanks Rex. we're in a much better place today. your insights were extremely valuable and we're now confident handlebars.net is doing its job as expected. your technology has allowed us to share templates between the front and back end which gives us great flexibility without duplication. 

John Robert Allan
john@johnrobertallan.com
+1.778.837.5218

**Comment by rexm** [comment]
Changes to the `PathBinder` to return an `UndefinedBindingResult` instead of throwing an exception has been included in NuGet 1.3.5. Thanks for the collaboration!
